File: examples/java/src/main/java/com/coinbase/cdp/examples/enduser/SendUserOperation.java
package com.coinbase.cdp.examples.enduser;

import com.coinbase.cdp.CdpClient;
import com.coinbase.cdp.examples.utils.EnvLoader;
import com.coinbase.cdp.openapi.model.EvmCall;
import com.coinbase.cdp.openapi.model.EvmUserOperationNetwork;
import com.coinbase.cdp.openapi.model.SendUserOperationWithEndUserAccountRequest;
import java.util.List;

/**
 * Example: Send a user operation on behalf of an end user.
 *
 * <p>This example requires the end user to have an active delegation on their account that allows
 * the developer to sign on their behalf.
 *
 * <p>User operations are sent through a smart account and can batch multiple calls together. This
 * example uses the CDP paymaster to sponsor gas fees.
 *
 * <p>Usage: ./gradlew runSendUserOperation --args="&lt;userId&gt;"
 */
public class SendUserOperation {

  public static void main(String[] args) throws Exception {
    if (args.length < 1) {
      System.err.println("Usage: SendUserOperation <userId>");
      System.exit(1);
    }

    String userId = args[0];
    EnvLoader.load();

    try (CdpClient cdp = CdpClient.create()) {
      // Get the end user to find their smart account address
      var endUser = cdp.endUser().getEndUser(userId);

      if (endUser.getEvmSmartAccountObjects().isEmpty()) {
        System.err.println("End user has no EVM smart accounts.");
        System.exit(1);
      }

      String address = endUser.getEvmSmartAccountObjects().get(0).getAddress();
      System.out.println("Sending user operation for user: " + userId);
      System.out.println("  Smart account address: " + address);
      System.out.println();

      var response =
          cdp.endUser()
              .sendUserOperation(
                  userId,
                  address,
                  new SendUserOperationWithEndUserAccountRequest()
                      .network(EvmUserOperationNetwork.BASE_SEPOLIA)
                      .calls(
                          List.of(
                              new EvmCall()
                                  .to(
                                      "0x0000000000000000000000000000000000000000")
                                  .value("0")
                                  .data("0x")))
                      .useCdpPaymaster(true));

      System.out.println("User operation sent successfully!");
      System.out.println("  User operation hash: " + response.getUserOpHash());
      System.out.println("  Status: " + response.getStatus());
    }
  }
}

File: examples/python/end_user/send_user_operation.py
# Usage: uv run python end_user/send_user_operation.py <user_id>
# Note: This example requires the end user to have an active delegation on their account
# that allows the developer to sign on their behalf.

import asyncio
import sys

from cdp import CdpClient
from dotenv import load_dotenv

load_dotenv()


async def main():
    if len(sys.argv) < 2:
        print("Usage: uv run python end_user/send_user_operation.py <user_id>")
        sys.exit(1)

    user_id = sys.argv[1]
    network = "base-sepolia"

    # Example calls for a user operation (a simple 0 ETH transfer).
    calls = [
        {
            "to": "0x0000000000000000000000000000000000000000",
            "value": "0",
            "data": "0x",
        }
    ]

    async with CdpClient() as cdp:
        try:
            # Get the end user to retrieve their EVM account address.
            end_user = await cdp.end_user.get_end_user(user_id=user_id)
            address = end_user.evm_accounts[0].address

            # Send a user operation using the client method.
            result = await cdp.end_user.send_user_operation(
                user_id=user_id,
                address=address,
                network=network,
                calls=calls,
                use_cdp_paymaster=True,
            )

            print("User op hash (client method):", result.user_op_hash)

            # Alternatively, send a user operation using the account method.
            # result = await end_user.send_user_operation(
            #     network=network,
            #     calls=calls,
            #     address=address,
            #     use_cdp_paymaster=True,
            # )

            # print("User op hash (account method):", result.user_op_hash)

        except Exception as e:
            print(f"Error sending user operation: {e}")
            raise e


asyncio.run(main())
